### PR TITLE
refactor: implement p2p paymentRepository

### PR DESCRIPTION
**Pull request by RD418**
This PR introduces full support for **peer-to-peer (P2P) marketplace transactions**, completing the feature requested in Issue #470.

⚠️ FOR REVIEWER: Please, correct me if I m wrong, but duplicated code comes from the "android.util.Log.d" lines I added, not from actual code duplication.

### **1. Updated Webhook Logic**
- Webhook now handles `payment.type = "marketplace"`.
- Implements **ticket reservation** during a buy request to prevent race conditions.
- On successful payment:
  - Creates the P2P transaction.
  - Locks the ticket.
  - Transfers ownership from seller → buyer.
- Adds extensive structured logging to simplify debugging.

### **2. New Cloud Function: `createMarketplacePaymentIntent`**
- Atomically creates a marketplace transaction before calling Stripe.
- Ensures strong consistency and prevents double purchases.
- Handles:
  - Payment intent creation
  - Stripe API errors & aborted flows
  - Invalid cases (e.g., `buyerId == sellerId`)
- Designed to be fully server-trusted and resistant to concurrency issues.

### **3. Repository Update**
- Added a repository function to call `createMarketplacePaymentIntent` from the client.
- ViewModels can now trigger P2P transaction flows cleanly and safely.

---

##  Result / Goal
This PR **enables secure, server-side P2P transactions** within the app, including racing-condition-safe ticket reservations and Stripe `stripeId` validation.  
The new marketplace logic runs fully separate from the existing direct-to-organization payment flow, improving reliability and expanding product capabilities.

Note: PR was structured by AI. The writing of tests was made by AI, the interface signature of the function was made by AI.

### CODE REVIEW

File: app/src/main/java/ch/onepass/onepass/model/payment/PaymentRepository.kt [hunk 1]
```diff
@@ -1,11 +1,42 @@
...
+   * ticket to prevent race conditions.
+   *
+   * @param ticketId The ID of the listed ticket to purchase.
+   * @param description Optional description for the payment.
+   * @return PaymentIntentResponse containing the client secret for the payment sheet.
+   */
+  suspend fun createMarketplacePaymentIntent(
+      ticketId: String,
+      description: String? = null
+  ): Result<MarketplacePaymentIntentResponse>
```
Comment (M-e-d-A) [inline app/src/main/java/ch/onepass/onepass/model/payment/PaymentRepository.kt]: **Question:** The KDoc says "This reserves the ticket to prevent race conditions" but doesn't document:
1. How long the reservation lasts (5 minutes based on the implementation)
2. What happens if the user never completes payment (does it auto-expire?)
3. Whether the client needs to manually cancel the reservation

Add this critical information to the documentation:

```kotlin
/**
 * Creates a payment intent for purchasing a listed ticket from the marketplace.
 * 
 * This function atomically reserves the ticket for 5 minutes to prevent race conditions.
 * The reservation will automatically expire if payment is not completed. Call
 * [cancelMarketplaceReservation] to manually release the reservation if the user
 * abandons the payment flow before the timeout.
 * 
 * @param ticketId The ID of the listed ticket to purchase
 * @param description Optional description for the payment
 * @return PaymentIntentResponse containing the client secret and ticket details
 * @throws Exception if ticket is not available, already reserved, or user is not authenticated
 */
```

File: app/src/main/java/ch/onepass/onepass/model/payment/PaymentRepositoryFirebase.kt
Comment (M-e-d-A) [inline app/src/main/java/ch/onepass/onepass/model/payment/PaymentRepositoryFirebase.kt]: **Suggestion:** The entire `createMarketplacePaymentIntent` function has significant code duplication with the existing `createPaymentIntent` function:

- Auth verification and token retrieval (duplicated)
- Stripe customer creation/retrieval (duplicated)  
- Error handling patterns (duplicated)

Extract common logic into private helper functions:

```kotlin
private suspend fun ensureAuthenticatedWithToken(
    errorMessage: String
): Result<Pair<FirebaseUser, String>> {
    val currentUser = auth.currentUser ?: return Result.failure(Exception(errorMessage))
    val token = try {
        currentUser.getIdToken(true).await().token
    } catch (e: Exception) {
        return Result.failure(Exception(errorMessage))
    }
    return if (token != null) {
        Result.success(currentUser to token)
    } else {
        Result.failure(Exception(errorMessage))
    }
}

private suspend fun ensureStripeCustomer(userId: String, userEmail: String?, userName: String?): String {
    // ... customer creation/retrieval logic
}
```

Then both functions can call these helpers, reducing duplication and improving maintainability.

File: app/src/test/java/ch/onepass/onepass/model/payment/PaymentRepositoryFirebaseTest.kt
Comment (M-e-d-A) [inline app/src/test/java/ch/onepass/onepass/model/payment/PaymentRepositoryFirebaseTest.kt]: The test coverage for createMarketplacePaymentIntent is comprehensive - great work! However, I notice some missing test cases:

Concurrent reservations: What happens if two users try to reserve the same ticket simultaneously?
Expired reservation override: What if user A's reservation expired and user B tries to reserve?
Price mismatch: What if the ticket price changes between viewing and purchasing?
Network timeout: What if the Stripe API call times out?
Partial response: What if Stripe returns a payment intent but Firestore write fails?

Consider adding these edge case tests, especially for concurrent access since that's a key feature of the reservation system.

File: functions/src/stripe/createMarketplacePaymentIntent.ts [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+);
+
+/**
+ * Cancels a marketplace ticket reservation.
+ * Called when the user abandons the payment flow.
+ *
+ * @param data - Request containing ticketId
+ * @param context - Firebase auth context
+ */
+export const cancelMarketplaceReservation = functions.https.onCall(
```
Comment (M-e-d-A) [inline functions/src/stripe/createMarketplacePaymentIntent.ts]: `cancelMarketplaceReservation` function looks good, but there's no check for whether the reservation has expired. Consider:

```typescript
const now = admin.firestore.Timestamp.now();

// ... in transaction:
if (ticketData?.reservedBy !== userId) {
    // Allow canceling expired reservations (idempotent)
    const reservedUntil = ticketData?.reservedUntil;
    if (reservedUntil && reservedUntil.toMillis() < now.toMillis()) {
        console.log(`Reservation for ticket ${ticketId} already expired`);
        return {success: true, reason: "already_expired"};
    }
    
    throw new functions.https.HttpsError(
        "permission-denied",
        "You can only cancel your own reservation"
    );
}
```

File: functions/src/stripe/createMarketplacePaymentIntent.ts [hunk 3]
```diff
@@ -0,0 +1,378 @@
+/**
+ * Creates a Stripe PaymentIntent for marketplace ticket purchases (P2P)
+ * 
+ * This function handles the complex flow of:
+ * 1. Validating the ticket is listed and available
+ * 2. Reserving the ticket to prevent race conditions
+ * 3. Creating a Stripe PaymentIntent
+ * 4. Storing the marketplace transaction record
+ */
```
[the remark this answers is not in the document]
Comment (RD418) [reply functions/src/stripe/createMarketplacePaymentIntent.ts]: Thanks

File: functions/src/stripe/webhooks.ts [hunk 4]
```diff
@@ -361,6 +284,234 @@ app.post("*", async (req: express.Request, res: express.Response) => {
...
+  });
+
+  console.log(`✅ Marketplace purchase completed: ${paymentIntent.id}`);
+}
+
+/**
+ * Handles successful event ticket purchase (original ticket creation flow).
+ * Creates new tickets and updates event availability.
+ */
+async function handleEventTicketPurchaseSuccess(
```
[the remark this answers is not in the document]
Comment (RD418) [reply functions/src/stripe/webhooks.ts]: Thanks

[the remark this answers is not in the document]
Comment (RD418) [reply functions/src/stripe/webhooks.ts]: So In general I think your issues will be addressed in the future, since they are not fatal. I will add documentation & modularize the code for visiblity. @M-e-d-A.

File: app/src/main/java/ch/onepass/onepass/model/payment/PaymentRepository.kt [hunk 5]
```diff
@@ -1,11 +1,42 @@
...
 
 interface PaymentRepository {
+  /** Creates a payment intent for purchasing event tickets directly from an event. */
   suspend fun createPaymentIntent(
       amount: Long,
       eventId: String,
       ticketTypeId: String? = null,
       quantity: Int = 1,
       description: String? = null
   ): Result<PaymentIntentResponse>
```
[the remark this answers is not in the document]
Comment (M-e-d-A) [reply app/src/main/java/ch/onepass/onepass/model/payment/PaymentRepository.kt]: Great!

File: app/src/main/java/ch/onepass/onepass/model/payment/PaymentRepositoryFirebase.kt
[the remark this answers is not in the document]
Comment (M-e-d-A) [reply app/src/main/java/ch/onepass/onepass/model/payment/PaymentRepositoryFirebase.kt]: As for the previous coment, noted!

File: functions/src/stripe/createMarketplacePaymentIntent.ts [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+  sellerId: string;
+  buyerId: string;
+  amount: number;
+  currency: string;
+  status: string;
+  type: "marketplace";
+  reservedUntil: admin.firestore.Timestamp;
+  createdAt: admin.firestore.FieldValue;
+  updatedAt: admin.firestore.FieldValue;
+}
```
[the remark this answers is not in the document]
Comment (M-e-d-A) [reply functions/src/stripe/createMarketplacePaymentIntent.ts]: Noted!

File: functions/src/stripe/createMarketplacePaymentIntent.ts (hunk 2 above)
Comment (M-e-d-A) [reply functions/src/stripe/createMarketplacePaymentIntent.ts]: Ok, noted!

File: app/src/test/java/ch/onepass/onepass/model/payment/PaymentRepositoryFirebaseTest.kt
Comment (RD418) [reply app/src/test/java/ch/onepass/onepass/model/payment/PaymentRepositoryFirebaseTest.kt]: - As mentionned above and in the code. Tickets cannot be reserved simultaenously. Reservations will necessarly pick one user, even in the odd situation where two are doing it at the exact time. 
- The user B reserves the spot then.
- it fails.
- It does not go through firestore, I don't get this question. On Success, the flow goes to the user directly.

Comment (M-e-d-A) [reply app/src/test/java/ch/onepass/onepass/model/payment/PaymentRepositoryFirebaseTest.kt]: Understood!

**Review by M-e-d-A — APPROVED** [review]
Thanks for the detailed answers, I see that most of the concerns have been addresed, especially the commitiment towards improvement is appreciated. Please make sure to update this thread when the documentation and code changes are pushed @RD418.
